lib/core/src/sdk.rs

```diff
@@ -2164,136 +2196,185 @@ impl LiquidSdk {
 
     /// This method fetches the chain tx data (onchain and mempool) using LWK. For every wallet tx,
     /// it inserts or updates a corresponding entry in our Payments table.
-    async fn sync_payments_with_chain_data(&self, with_scan: bool) -> Result<()> {
-        let payments_before_sync: HashMap<String, Payment> = self
-            .list_payments(&ListPaymentsRequest::default())
-            .await?
-            .into_iter()
-            .flat_map(|payment| {
-                // Index payments by both tx_id (lockup/claim) and refund_tx_id
-                let mut res = vec![];
-                if let Some(tx_id) = payment.tx_id.clone() {
-                    res.push((tx_id, payment.clone()));
-                }
-                if let Some(refund_tx_id) = payment.get_refund_tx_id() {
-                    res.push((refund_tx_id, payment));
+    async fn sync_payments_with_chain_data(&self, partial_sync: bool) -> Result<()> {
+        self.onchain_wallet.full_scan().await?;
+
+        let mut tx_map = self.onchain_wallet.transactions_by_tx_id().await?;
+        let swaps_list = self.get_monitored_swaps_list(partial_sync).await?;
+        let recovered_onchain_data = self
+            .recover_from_onchain(
+                TxMap::from_raw_tx_map(tx_map.clone()),
+                swaps_list,
+                partial_sync,
+            )
+            .await?;
+
+        // Loop over the recovered chain data for monitored receive swaps
+        for (swap_id, receive_data) in recovered_onchain_data.receive {
+            if let Some(to_state) = receive_data.derive_partial_state() {
+                let lockup_tx_id = receive_data.lockup_tx_id.map(|h| h.txid.to_string());
+                let claim_tx_id = receive_data.claim_tx_id.clone().map(|h| h.txid.to_string());
+                let mrh_tx_id = receive_data.mrh_tx_id.clone().map(|h| h.txid.to_string());
+                _ = self
+                    .receive_swap_handler
+                    .update_swap_info(
```

Comment:
Will this send trigger a payment event? Previously the event was triggered only on change.

Reply:
I think they are repeated anyway in some cases. I seem to remember seeing multiple PaymentPending or PaymentWaitingConfirmation events. Probably because its valid to update Pending to Pending even if the details are the same.

Reply:
I mean also on the update_swap_info occurences bellow we don't seem to be checking if it is a really state transition so it looks like on every sync we will get payment events.
Previousely the code was checking for every associated claim if it really transitioned to from mempool to confirmed (completed) or just intruduced as undonfirmed (pending) before emitting events.